botorch/acquisition/max_value_entropy_search.py

```diff
@@ -313,6 +320,206 @@ def _compute_information_gain(
+        # batch_shape x 1
+        check_no_nans(variance_noisy)
+        variance_noiseless = posterior_noiseless.variance.clamp_min(CLAMP_LB).squeeze(
+            -1
+        )
+        # batch_shape x 1
+        check_no_nans(variance_noisy)
+        stdv = variance_noiseless.sqrt()
+
+        # define normal distribution to compute cdf and pdf
+        normal = torch.distributions.Normal(
+            torch.zeros(1, device=X.device, dtype=X.dtype),
+            torch.ones(1, device=X.device, dtype=X.dtype),
+        )
+
+        # prepare max value quantities required by GIBBON
+        mvs = torch.transpose(self.posterior_max_values, 0, 1)
+        # 1 x s_M
+        normalized_mvs = (mvs - mean) / stdv
+        # batch_shape x s_M
+        cdf_mvs = normal.cdf(normalized_mvs).clamp_min(CLAMP_LB)
+        pdf_mvs = torch.exp(normal.log_prob(normalized_mvs))
+        ratio = pdf_mvs / cdf_mvs
+        check_no_nans(ratio)
+
+        # prepare variance ratio
+        rhos_squared = variance_noiseless / variance_noisy
+        # batch_shape x 1
+        check_no_nans(rhos_squared)
+
+        # build acquisition function
+        inner_log = 1 - rhos_squared * ratio * (normalized_mvs + ratio)
+        acq = -0.5 * inner_log.clamp_min(CLAMP_LB).log()
+        # average over Gumbel samples
+        acq = acq.mean(dim=1)
+
+        if self.X_pending is not None:
+            # for q>1 require repulsion terms r_i = log |C_i| for the predictive
+            # correlation matricies between each candidate point in X and
+            # the m current batch elements in X_pending.
+
+            # Each predictive covariance matrix can be expressed as
+            # V_i = [[v_i, A_i], [A_i,B]] for a shared m x m tensor B.
+            # So we can efficientely calculate |V_i| using formula for determinant of block matricies
+            v = variance_noisy
+            full_covariance = self.model(
+                torch.cat([X.squeeze(1), self.X_pending], dim=0)
+            )
+            # only evaluate required cross covariance elements
+            A = (
+                full_covariance.lazy_covariance_matrix[: len(X), len(X) :]
+                .evaluate()
+                .unsqueeze(1)
+            )
+            # batch_shape x 1
+            B = self.model.posterior(
+                self.X_pending, observation_noise=True
+            ).mvn.covariance_matrix
+            B_inv = torch.inverse(B)
+            covariance_determinant = v - torch.matmul(
+                torch.matmul(A, B_inv), A.transpose(1, 2)
+            ).squeeze(-1)
+            covariance_determinant = covariance_determinant * torch.det(B)
+            # batch_shape x 1
+
+            # Take logs and convert covariances to correlations.
+            r = covariance_determinant.log() - v.log() - torch.diagonal(B).log().sum()
+            r = 0.5 * r.squeeze(1)
+            return acq + r
+
+        else:
+            # for q=1, no replusion term required
+            return acq
```

Comment:
> This interface can be possible made more general by defining a MaxValueEntropyBase() class as suggested by @Balandat and include a H1() abstract method inside, since so far almost all MES related acquisition function that I know differs only in how H1() part is estimated/approximated.

> I also strongly recommend to use qLowerBoundMaxValueEntropy or q...MaxValueEntropy to conform with the naming convention, which is also much easier for user to know about what this acquisition is about without needing to read the paper.

This makes a lot of sense, then we can share all of the boilerplate code and allow implementing different variants directly by just subclassing and overwriting a single method. I would suggest we
1. Substantially improve `qMaxValueEntropy` and build a class that takes care of the `H0` computation, and also provides a general interface for sampling posterior max values (we have discretization-based and Gumbel variants, we could easily also implement optimizing RFF draws or using decoupled sampling approaches as in @j-wilson's recent work. I also have some ideas for how to do this exactly in a scalable fashion - cc @vishwakftw, @dme65).
2. Make sure this `qMaxValueEntropy` still retains the current behavior (possibly modulo some of the limitations that raise errors)
3. Implement `qLowerBoundMaxValueEntropy` and any other / future modification of it as subclassess of `qMaxValueEntropy`.

We can then do the same thing for `qMultiFidelityMaxValueEntropy`.

@henrymoss not sure if this is something that you've singed up for - I'm happy to take on this larger refactor if you feel that that is outside the scope of your PR.

Reply:
Thanks so much for all the comments, everyone.

I will have a go at making these changes today!

Reply:
I have started abstracting the classes. Thanks for all your helpful comments.

 @liangshi7 , we found joint allocation of batches to perform poorly and so only support greedy allocation (i.e the same as your current batch MES offerings). 

You are right that we would need X_pending to calculate the actual value of information  gain. However, as we are building batches greedily, we only need the relative improvements in information gain from adding in a new point to the batch.

Reply:
Also @liangshi7 , we found little to no benefit from using MC integral approximation rather than moment-matching when building batches

Reply:
>  we found joint allocation of batches to perform poorly and so only support greedy allocation (i.e the same as your current batch MES offerings).

interesting, do you have some benchmark studies on synthetic functions with q-batch queries? 

When we worked on LBMES with MC integration, we did some comparison to MES and KG on synthetic functions (below is for Hartmann6 and q-batch=4, q-MES-LB-join is our LBMES with MC integrations), the performance on regrets was pretty competitive (only slighly worse) with KG and MES, but with a significant reduction of computing time similar to what you show in your paper. If helpful, we could open-source our internal LBMES so that you can do some comparative studies. 

![image](https://user-images.githubusercontent.com/13260879/109794109-868b4980-7be3-11eb-8a4e-16981b59eb58.png)

![image](https://user-images.githubusercontent.com/13260879/109794416-d9650100-7be3-11eb-826b-57493c2ad4e5.png)


> we found little to no benefit from using MC integral approximation rather than moment-matching when building batches

This is possible given that the ESG distribution is probably very similar to a Gaussian anyway. One way to show this is to plot the approximation error of the differential entropy using moment-matching relative to using a high-order numerical integration.